powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/repository/ActivationRepository.java

```diff
@@ -104,14 +104,18 @@
     ActivationRecordEntity findCreatedActivationWithoutLock(Long applicationId, String activationCode, Collection<ActivationStatus> states, Date currentTimestamp);
 
     /**
-     * Get count of activations identified by an activation code associated with given application.
+     * Get count of activations identified by an activation short ID associated with given application.
+     *
+     * The check for the first half of activation code is required for version 2.0 of PowerAuth crypto. In future the
+     * uniqueness check will be extended to whole activation code once version 2.0 of PowerAuth crypto is no longer
+     * supported.
      *
      * @param applicationId     Application ID
-     * @param activationCode    Activation code
+     * @param activationIdShort Activation ID short
      * @return Count of activations matching the search criteria
      */
-    @Query("SELECT COUNT(a) FROM ActivationRecordEntity a WHERE a.application.id = ?1 AND a.activationCode = ?2")
-    Long getActivationCountByActivationCode(Long applicationId, String activationCode);
+    @Query("SELECT COUNT(a) FROM ActivationRecordEntity a WHERE a.application.id = ?1 AND a.activationCode LIKE ?2%")
+    Long getActivationCountByActivationIdShort(Long applicationId, String activationIdShort);
```

Comment:
I am not sure about this change. Crypto 3.0 does not have a "short activation ID", it only has an activation code. Therefore, I believe that the Java code should be written to cover Crypto 3.0. To check strong uniqueness, I would prefer to modify the HQL by using [`substring`](http://www.hplsql.org/substring) in the query...

Reply:
We still support crypto v2 which contains short activation ID as a parameter.

See:
https://github.com/wultra/powerauth-server/blob/develop/powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v2/ActivationServiceBehavior.java#L194

https://github.com/wultra/powerauth-server/blob/develop/powerauth-java-server/src/main/resources/xsd/PowerAuth-2.0.xsd#L34

We will remove this method when crypto v2 is deprecated.

As for using `substring`, I would not recommend it, because it would degrade DB performance. When `LIKE ?2%` is used, index on activation code is leveraged in each database PowerAuth supports. In case of `substring` we would need to define a special index on the function, which can be tricky in some databases.